lightgbm/src/main/scala/com/microsoft/azure/synapse/ml/lightgbm/booster/LightGBMBooster.scala

```diff
@@ -264,13 +264,15 @@ class LightGBMBooster(val trainDataset: Option[LightGBMDataset] = None, val para
   }
 
   /** Saves the booster to string representation.
+    * @param saveAllIterations Whether to save all iterations, or up to the best one.
     * @return The serialized string representation of the Booster.
     */
-  def saveToString(): String = {
+  def saveToString(saveAllIterations: Boolean): String = {
       val bufferLength = LightGBMConstants.DefaultBufferLength
       val bufferOutLengthPtr = lightgbmlib.new_int64_tp()
+      val lastIteration: Int = if (saveAllIterations) -1 else bestIteration
       lightgbmlib.LGBM_BoosterSaveModelToStringSWIG(boosterHandler.boosterPtr,
-        0, -1, 0, bufferLength, bufferOutLengthPtr)
+        0, lastIteration, 0, bufferLength, bufferOutLengthPtr)
```

Comment:
So from what I see we are fixing an issue where the early-stopping round is being ignored, where saving the model.

It may be helpful to see what LightGBM itself does. In LightGBM itself, in normal circumstances in [its own training loop at least in the `GBDT` implementation of its `Booster` interface],(https://github.com/microsoft/LightGBM/blob/6b56a90cd1324a6dbac2afa0a352c9355b0dc3cf/src/boosting/gbdt.cpp#L266-L285) (which we do not use) it works by checking the early stopping criteria from time to time and if reached [prunes the model of the excess trees](https://github.com/microsoft/LightGBM/blob/6b56a90cd1324a6dbac2afa0a352c9355b0dc3cf/src/boosting/gbdt.cpp#L481-L483).

I believe we are bypassing that logic in SynapseML's wrapper -- that is, we do not prune LightGBM's internal implementation, as LightGBM itself does. Otherwise presumably this bug, to the extent that I understand that there is a bug, would not exist. Instead, what we are doing is maintaining the full model, and are implementing early stopping ourselves, from scratch, for what I presume are reasonable reasons. (Though for my own sake in understanding the wrapper I would not object if this was elucidated. It is far from obvious to me that if I were writing the wrapper, I wouldn't use LightGBM's own facilities for doing this, as opposed to implementing them myself.)

One issue that has cropped up, apparently, is that people notice that when saving the model, all the iterations are saved to the file. To fix this issue, we are passing in that `lastIteration` into this where we save the textual representation of the tree to a file. Are we quite certain that this is the only place where this "ignoring early stopping" bug manifests itself?

Most seriously, it is very unclear to me when doing our raw model score calculation, whether the best iteration is taken into account. The booster's [`score`](https://github.com/microsoft/SynapseML/blob/6b9dadb80bb6ecbef69992a09a448f1afa13a71f/lightgbm/src/main/scala/com/microsoft/azure/synapse/ml/lightgbm/booster/LightGBMBooster.scala#L400) method, if I trace through all code paths, does not appear to do that.

Have we considered doing what LightGBM appears to do based on my reading above, and pruning the native model itself, as a final step of the training process?

Reply:
Ah, maybe it does... I see that we are changing `numIterations` at the same time we change `bestIteration` in the call to `setBestIteration`.

https://github.com/microsoft/SynapseML/blob/6b9dadb80bb6ecbef69992a09a448f1afa13a71f/lightgbm/src/main/scala/com/microsoft/azure/synapse/ml/lightgbm/booster/LightGBMBooster.scala#L444-L447

This strikes me as a somewhat baffling choice. It appears to be just flat out wrong -- the number of iterations is set to a value that it wasn't. Why have two distinct fields at all then?